mqtt/mqtt-bridge/src/persist/waking_state/ring_buffer/mod.rs

```diff
@@ -225,8 +225,22 @@ impl StreamWakeableState for RingBuffer {
         let result = load_block_header(&mut self.file, start, block_size, self.max_file_size);
         if let Ok(block_header) = result {
             let BlockVersion::Version1(inner) = block_header.inner();
+            // It is possible that we end up writing a smaller block on wrap-around,
+            // where the file pointers will end up being in the middle of random data.
+            // To avoid this, we need to check if we have a block of data, if we do,
+            // we will need to just allow the write. Above cases will prevent us from
+            // accidentally corrupting.
+            let mut allow_write = false;
+            let bin_block = bincode::serialize(inner)?;
+            if bin_block.iter().map(|&x| u64::from(x)).sum::<u64>() != 0
+                && inner.hint() != BLOCK_HINT
+            {
+                // This is a bad block, where we wrote something smaller
+                // and this is okay, we just need to allow the write.
+                allow_write = true;
+            }
             let should_not_overwrite = inner.should_not_overwrite();
-            if should_not_overwrite {
+            if !allow_write && should_not_overwrite {
```

Comment:
I'm not sure I completely understand the issue. Is it that you were able to deserialize block header from junk data and `should_not_overwrite` gave you false positive?

Reply:
Also, I'm more and more skeptical about this check at all. I understand we still need `should_not_overwrite` flag for crash recovery, but do we need that for normal work? can we just rely on pointers?

Reply:
The issue is that yes, we deserialized from junk data, which is okay since we had data there and we are now writing new data which does not cover the whole of previous data. We can get whatever value happens to map over that region of data, whether it is true or false. 

It can be that write goes faster than read. The `should_not_overwrite` flag prevents us from writing when we shouldn't. But, in the case of normal read and write operations without crash, we can likely get away with just the pointers. This would mean we can drop this check altogether, which should make this look nicer too

Reply:
Have removed this check entirely and it operates well without it :)